internal/kuberesource/mutators.go

```diff
@@ -5,42 +5,78 @@ package kuberesource
+const (
+	exposeServiceAnnotation      = "contrast.edgeless.systems/expose-service"
+	contrastRoleAnnotationKey    = "contrast.edgeless.systems/pod-role"
+	skipInitializerAnnotationKey = "contrast.edgeless.systems/skip-initializer"
+)
 
 // AddInitializer adds an initializer and its shared volume to the resource.
 //
 // If the resource does not contain a PodSpec, this function does nothing.
-// This function is not idempotent.
+// This function is idempotent.
 func AddInitializer(
 	resource any,
 	initializer *applycorev1.ContainerApplyConfiguration,
-) any {
-	return MapPodSpec(resource, func(spec *applycorev1.PodSpecApplyConfiguration) *applycorev1.PodSpecApplyConfiguration {
-		// Add the initializer as an init container.
-		spec.WithInitContainers(
-			initializer,
-		)
+) (res any, err error) {
+	res = MapPodSpec(resource, func(meta *applymetav1.ObjectMetaApplyConfiguration, spec *applycorev1.PodSpecApplyConfiguration) *applycorev1.PodSpecApplyConfiguration {
+		if meta.Annotations[skipInitializerAnnotationKey] == "true" {
+			return spec
+		}
+		if spec.RuntimeClassName == nil || !strings.HasPrefix(*spec.RuntimeClassName, "contrast-cc") {
+			return spec
+		}
+
+		// Remove already existing init containers with unique initializer name.
+		spec.InitContainers = slices.DeleteFunc(spec.InitContainers, func(c applycorev1.ContainerApplyConfiguration) bool {
+			return *c.Name == *initializer.Name
+		})
+		// Add the initializer as first init container.
+		spec.InitContainers = append([]applycorev1.ContainerApplyConfiguration{*initializer}, spec.InitContainers...)
 		if len(initializer.VolumeMounts) < 1 {
 			return spec
 		}
-		// Create the volume written by the initializer.
-		spec.WithVolumes(Volume().
-			WithName(*initializer.VolumeMounts[0].Name).
-			WithEmptyDir(EmptyDirVolumeSource().Inner()),
-		)
+
+		// Existing volume with unique name has to be of type EmptyDir.
+		var volumeExists bool
+		for _, volume := range spec.Volumes {
+			if *volume.Name == *initializer.VolumeMounts[0].Name {
+				volumeExists = true
+				if volume.EmptyDir == nil {
+					err = fmt.Errorf("volume %s has to be of type EmptyDir", *volume.Name)
+					return nil
+				}
+			}
+		}
+		// Create the volume written by the initializer if it not already exists.
+		if !volumeExists {
+			spec.WithVolumes(Volume().
+				WithName(*initializer.VolumeMounts[0].Name).
+				WithEmptyDir(EmptyDirVolumeSource().Inner()),
+			)
+		}
+
+		// Remove already existing volume mounts on the worker container with unique volume mount name.
+		spec.Containers[0].VolumeMounts = slices.DeleteFunc(spec.Containers[0].VolumeMounts, func(v applycorev1.VolumeMountApplyConfiguration) bool {
+			return *v.Name == *initializer.VolumeMounts[0].Name
+		})
+
 		// Add the volume mount written by the initializer to the worker container.
 		spec.Containers[0].WithVolumeMounts(VolumeMount().
 			WithName(*initializer.VolumeMounts[0].Name).
 			WithMountPath(*initializer.VolumeMounts[0].MountPath))
```

Comment:
This needs to be done for all containers, right? E.g., when adding the service mesh but also dealing with some part of TLS manually...